target/linux/airoha/files/drivers/pinctrl/pinctrl-airoha.c

```diff
@@ -2118,9 +2136,155 @@ static int airoha_pinctrl_add_gpiochip(struct airoha_pinctrl *pinctrl,
 	return devm_gpiochip_add_data(dev, chip, pinctrl);
 }
 
+static struct airoha_gpiochip *chip_to_ctrl(struct gpio_chip *chip)
+{
+	return container_of(chip, struct airoha_gpiochip, chip);
+}
+
+
+static irqreturn_t airoha_gpio_irq_handler(int irq, void *data)
+{
+	struct gpio_chip *chip = data;
+	struct airoha_gpiochip *ctrl = chip_to_ctrl(chip);
+
+	unsigned long pending;
+	int bit;
+
+	pending = ioread32(ctrl->irqstatus);
+	if(!pending)
+		return IRQ_NONE;
+
+	for_each_set_bit(bit, &pending, AIROHA_GPIO_MAX) {
+		u32 map = irq_find_mapping(chip->irq.domain, bit);
+
+		generic_handle_irq(map);
+		iowrite32(BIT(bit), ctrl->irqstatus);
+	}
+
+	return IRQ_HANDLED;
+}
+
+static void airoha_gpio_irq_unmask(struct irq_data *data)
+{
+
+	struct airoha_gpiochip *ctrl = irq_data_get_irq_chip_data(data);
+	int pin = data->hwirq;
+	unsigned long flags;
+	u32 tmp;
+	u32 mask = BIT((pin % 16) * 2) | (BIT((pin % 16) * 2) << 1);
+
+	printk("airoha_gpio_irq_unmask: pin = %d mask = %x\n", pin, mask);
```

Comment:
@merbanan can you please get rid of printk?

Reply:
Ack